pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -2017,6 +2027,25 @@ private void handleLastMessageIdFromCompactedLedger(PersistentTopic persistentTo
         });
     }
 
+    private int calculateTheLastBatchIndexInBatch(MessageMetadata metadata, ByteBuf payload) throws IOException {
+        int batchSize = metadata.getNumMessagesInBatch();
+        if (batchSize <= 1){
+            return -1;
+        }
+        SingleMessageMetadata singleMessageMetadata = new SingleMessageMetadata();
+        int lastBatchIndexInBatch = -1;
+        for (int i = 0; i < batchSize; i++){
+            ByteBuf singleMessagePayload =
+                    Commands.deSerializeSingleMessageInBatch(payload, singleMessageMetadata, i, batchSize);
+            singleMessagePayload.release();
+            if (singleMessageMetadata.isCompactedOut()){
+                continue;
```

Comment:
Is it better to judge in reverse order? if the lastBatchIndex is not `CompactedOut` can return directly

Reply:
Hi Bo
The constructor of batched message entry is `[single_msg_size_1, single_msg_content_1, single_msg_size_2, single_msg_content_2......]`, we can not extract the last internal message directly. same as https://github.com/apache/pulsar/pull/18877/files/b428afd7916b94542a20acf8f1f029bddcc0da81#r1050335022